### PR TITLE
Cache number of rows in files for count in file/s3/url/hdfs/azure functions

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Cache number of rows in files for count in file/s3/url/hdfs/azure functions. The cache can be enabled/disabled by setting `use_cache_for_count_from_files` (enabled by default). Continuation of https://github.com/ClickHouse/ClickHouse/pull/53637

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
This is an automated comment for commit c0bdd0e00bb8b6f189bd8d1dbad0e05ceff92c4b with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53692/c0bdd0e00bb8b6f189bd8d1dbad0e05ceff92c4b/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Storages/StorageFile.cpp [hunk 1]
```diff
@@ -1024,6 +1067,11 @@ class StorageFileSource : public ISource
             if (storage->use_table_fd)
                 finished_generate = true;
 
+            if (input_format && storage->format_name != "Distributed" && context->getSettingsRef().use_cache_for_count_from_files)
```
[the remark this answers is not in the document]
Comment (Avogar) [reply src/Storages/StorageFile.cpp:1070]: That's special case for storage file when we have the namfe of input format `Distributed`, it's not storage name. We have such checks in several places here

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 2]
```diff
@@ -2266,6 +2266,10 @@ std::optional<UInt64> InterpreterSelectQuery::getTrivialCount(UInt64 max_paralle
     auto & query = getSelectQuery();
     if (!query.prewhere() && !query.where() && !context->getCurrentTransaction())
     {
+        /// Some storages can optimize trivial count in read() method instead of totalRows() because it still can
+        /// require reading some data (but much faster than reading columns).
+        /// Set a special flag in query info so the storage will see it and optimize count in read() method.
+        query_info.optimize_trivial_count = optimize_trivial_count;
```
[the remark this answers is not in the document]
Comment (Avogar) [reply src/Interpreters/InterpreterSelectQuery.cpp:2272]: Sure, did it in https://github.com/ClickHouse/ClickHouse/pull/53637 right now

File: src/Processors/Formats/Impl/ArrowBlockInputFormat.h [hunk 3]
```diff
@@ -66,9 +66,16 @@ class ArrowSchemaReader : public ISchemaReader
...
 
+    std::optional<size_t> readNumberOrRows() override;
+
 private:
+    void initializeIfNeeded();
+
     bool stream;
     const FormatSettings format_settings;
+    std::shared_ptr<arrow::RecordBatchReader> stream_reader;
+    // The following fields are used only for Arrow format
```
[the remark this answers is not in the document]
Comment (Avogar) [reply src/Processors/Formats/Impl/ArrowBlockInputFormat.h]: The comment was copied from another code by mistake, removed

**Review by novikd — APPROVED** [review]

**Pull request merged**